### PR TITLE
[NAVAND-830] payment methods [DO-NOT-MERGE]

### PR DESCRIPTION
**Pull request by VysotskiVadim**

### CODE REVIEW

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java [hunk 1]
```diff
@@ -417,4 +482,26 @@ private DirectionsCriteria() {
   })
   public @interface ApproachesCriteria {
   }
+
+  /**
+   * Retention policy for the various payment methods.
```
Comment (VysotskiVadim) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:487]: Why do we call it a retention policy? It thought it means something different in the context of JVM 🤔

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java [hunk 2]
```diff
@@ -270,6 +270,71 @@ public final class DirectionsCriteria {
...
+
+  /***
+   * TODO: clarify.
+   */
+  public static final String PAYMENT_METHOD_CRYPTOCURRENCIES = "cryptocurrencies";
+
+  /***
+   * TODO: clarify.
+   */
+  public static final String PAYMENT_METHOD_APPS = "apps";
```
Comment (VysotskiVadim) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:336]: I will clarify exact description for each method.

**Comment by VysotskiVadim** [comment]
@LukasPaczos , @RingerJK , @dzinad , the PR isn't ready to be merged but it's ready for review 🙂

**Review by dzinad — APPROVED** [review]

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java [hunk 3]
```diff
@@ -768,6 +769,45 @@ public void emptyExcludeObjectsCleansUpExcludes() {
...
+
+    assertEquals("etc,general", routeOptions.paymentMethods());
+    assertEquals(paymentMethods, routeOptions.paymentMethodsList());
+    assertContains(query, "payment_methods=etc%2Cgeneral");
+  }
+
+  @Test
+  public void emptyPaymentMethodsList() {
+    RouteOptions routeOptions = routeOptions().toBuilder()
+      .paymentMethodsList(Collections.<String>emptyList())
```
[the remark this answers is not in the document]
Comment (VysotskiVadim) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:802]: Nope. I just felt confidence in case of null, because it's a default value. I didn't felt confidence in case of an empty list and added the test 🙂 
If you think it's important, I will add a test for the null case 👍

[the remark this answers is not in the document]
Comment (dzinad) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:802]: What will happen? Will there be a parameter with an empty string or no parameter at all? Since it's no parameter at all for an empty list I guess it will be the same for null so here I don't think it's necessary. I usually add tests for null and empty as a rule, just to be sure. No pressure tho. :)

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 4]
```diff
@@ -1978,6 +2000,29 @@ public Builder unrecognizedProperties(@Nullable Map<String, String> unrecognized
...
 
+    /***
+     * Enables filtering the route and lane by payment methods. Default is no filtering.
+     * Available payment methods match {@link DirectionsCriteria.PaymentMethodsCriteria}
+     * @param paymentMethods is a comma separated string where each value
+     *                       matches {@link DirectionsCriteria.PaymentMethodsCriteria}
+     * @return this builder
+     */
+    @NonNull
+    public abstract Builder paymentMethods(@Nullable String paymentMethods);
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:2011]: We're missing updates to `MapboxDirections` that use this new parameter (and corresponding request/response tests).

Comment (VysotskiVadim) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:2011]: Added. But I'm not sure about response tests in the `services-directions` module. They seem like a duplication of tests from the `services-dirctions-model` module, as they just check how a JSON test file transforms to model. WDYT?

**Comment by dzinad** [comment]
Fixed the conflicts, but the feature is temporarily removed from backend. Will place the ticket in "to track" and wait for the updates (targeting next week).

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java (hunk 1 above)
Comment (RingerJK) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/DirectionsCriteria.java:566]: I think it's legacy from the previous version, that we copy-and-paste whenever `Def`s are used. Feel free to change it

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/IntersectionLanes.java [hunk 5]
```diff
@@ -81,6 +80,16 @@ public static Builder builder() {
   @Nullable
   public abstract List<String> indications();
 
+  /*
+   * Available payment methods for the line.
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/IntersectionLanes.java:84]:
```suggestion
   * Available payment methods for the lane.
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/IntersectionLanes.java [hunk 6]
```diff
@@ -167,6 +176,14 @@ public abstract static class Builder extends DirectionsJsonObject.Builder<Builde
      */
     public abstract Builder indications(@Nullable List<String> indications);
 
+    /*
+     * Set available payment methods for the line.
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/IntersectionLanes.java:180]:
```suggestion
     * Set available payment methods for the lane.
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 7]
```diff
@@ -909,6 +909,29 @@ public List<Boolean> snappingIncludeStaticClosuresList() {
   @Nullable
   public abstract Boolean metadata();
 
+  /*
+   * Enables filtering the route and lane by payment methods. Default is no filtering.
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:913]:
```suggestion
   * Enables filtering the route and valid lanes by payment methods. Default is no filtering.
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 8]
```diff
@@ -909,6 +909,29 @@ public List<Boolean> snappingIncludeStaticClosuresList() {
...
+   * @return a comma separated string where each value
+   *         matches {@link DirectionsCriteria.PaymentMethodsCriteria}
+   */
+  @SuppressWarnings("checkstyle:javadocmethod")
+  @SerializedName("payment_methods")
+  @Nullable
+  public abstract String paymentMethods();
+
+  /*
+   * Enables filtering the route and lane by payment methods. Default is no filtering.
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:924]:
```suggestion
   * Enables filtering the route and valid lanes by payment methods. Default is no filtering.
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 9]
```diff
@@ -2090,6 +2114,31 @@ public Builder unrecognizedProperties(@Nullable Map<String, String> unrecognized
       return unrecognized(null);
     }
 
+    /*
+     * Enables filtering the route and lane by payment methods. Default is no filtering.
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:2118]:
```suggestion
     * Enables filtering the route and valid lanes by payment methods. Default is no filtering.
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 10]
```diff
@@ -2090,6 +2114,31 @@ public Builder unrecognizedProperties(@Nullable Map<String, String> unrecognized
...
+     * @param paymentMethods is a comma separated string where each value
+     *                       matches {@link DirectionsCriteria.PaymentMethodsCriteria}
+     * @return this builder
+     */
+    @SuppressWarnings("checkstyle:javadocmethod")
+    @NonNull
+    public abstract Builder paymentMethods(@Nullable String paymentMethods);
+
+    /*
+     * Enables filtering the route and lane by payment methods. Default is no filtering.
```
Comment (LukasPaczos) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:2129]:
```suggestion
     * Enables filtering the route and valid lanes by payment methods. Default is no filtering.
```

**Review by LukasPaczos — COMMENTED** [review]
LGTM but let's double-check the lane data validity and docs before merging, as discussed internally.
